### PR TITLE
server: fix a unstable test TestInitConnect

### LINKED ISSUE #30249: Unstable test found in CI TestInitConnect
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
[2021-11-30T02:29:26.743Z] --- FAIL: TestInitConnect (1.36s)
[2021-11-30T02:29:26.743Z]     dbtestkit.go:72:
[2021-11-30T02:29:26.743Z]              Error Trace:    dbtestkit.go:72
[2021-11-30T02:29:26.743Z]                                                      server_test.go:1939
[2021-11-30T02:29:26.743Z]                                                      server_test.go:117
[2021-11-30T02:29:26.743Z]                                                      server_test.go:1938
[2021-11-30T02:29:26.743Z]                                                      tidb_test.go:1107
[2021-11-30T02:29:26.743Z]              Error:          Received unexpected error:
[2021-11-30T02:29:26.743Z]                              commands out of sync. Did you run multiple statements at once?
[2021-11-30T02:29:26.743Z]              Test:           TestInitConnect
[2021-11-30T02:29:26.743Z]              Messages:       sql:SET GLOBAL init_connect="insert into test.ts VALUES (NOW());SET @a=1;", args:[]
[2021-11-30T02:29:26.743Z]     server_test.go:112:
[2021-11-30T02:29:26.743Z]              Error Trace:    server_test.go:112
[2021-11-30T02:29:26.743Z]                                                      panic.go:613
[2021-11-30T02:29:26.743Z]                                                      testing.go:741
[2021-11-30T02:29:26.743Z]                                                      dbtestkit.go:72
[2021-11-30T02:29:26.743Z]                                                      server_test.go:1939
[2021-11-30T02:29:26.743Z]                                                      server_test.go:117
[2021-11-30T02:29:26.743Z]                                                      server_test.go:1938
[2021-11-30T02:29:26.743Z]                                                      tidb_test.go:1107
[2021-11-30T02:29:26.743Z]              Error:          Received unexpected error:
[2021-11-30T02:29:26.743Z]                              bad connection
[2021-11-30T02:29:26.743Z]              Test:           TestInitConnect
```

### 2. What did you expect to see? (Required)

CI pass

### 3. What did you see instead (Required)

Sometimes fail ... I don't know how it happens

### 4. What is your TiDB version? (Required)

```
Head:     master executor: send a task with error to the resultCh when panic happen (#30214)
Merge:    origin/master executor: send a task with error to the resultCh when panic happen (#30214)
Tag:      v5.4.0-alpha (263)
```

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #30249

Problem Summary:
```
[2021-11-30T02:29:26.743Z] --- FAIL: TestInitConnect (1.36s)
[2021-11-30T02:29:26.743Z]     dbtestkit.go:72:
[2021-11-30T02:29:26.743Z]              Error Trace:    dbtestkit.go:72
[2021-11-30T02:29:26.743Z]                                                      server_test.go:1939
[2021-11-30T02:29:26.743Z]                                                      server_test.go:117
[2021-11-30T02:29:26.743Z]                                                      server_test.go:1938
[2021-11-30T02:29:26.743Z]                                                      tidb_test.go:1107
[2021-11-30T02:29:26.743Z]              Error:          Received unexpected error:
[2021-11-30T02:29:26.743Z]                              commands out of sync. Did you run multiple statements at once?
[2021-11-30T02:29:26.743Z]              Test:           TestInitConnect
[2021-11-30T02:29:26.743Z]              Messages:       sql:SET GLOBAL init_connect="insert into test.ts VALUES (NOW());SET @a=1;", args:[]
[2021-11-30T02:29:26.743Z]     server_test.go:112:
[2021-11-30T02:29:26.743Z]              Error Trace:    server_test.go:112
[2021-11-30T02:29:26.743Z]                                                      panic.go:613
[2021-11-30T02:29:26.743Z]                                                      testing.go:741
[2021-11-30T02:29:26.743Z]                                                      dbtestkit.go:72
[2021-11-30T02:29:26.743Z]                                                      server_test.go:1939
[2021-11-30T02:29:26.743Z]                                                      server_test.go:117
[2021-11-30T02:29:26.743Z]                                                      server_test.go:1938
[2021-11-30T02:29:26.743Z]                                                      tidb_test.go:1107
[2021-11-30T02:29:26.743Z]              Error:          Received unexpected error:
[2021-11-30T02:29:26.743Z]                              bad connection
[2021-11-30T02:29:26.743Z]              Test:           TestInitConnect
```

### What is changed and how it works?

> commands out of sync. Did you run multiple statements at once?

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by xhebox — APPROVED** [review]

**Review by djshow832 — APPROVED** [review]

**Comment by djshow832** [comment]
Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 60 near ""insert into test.ts VALUES (NOW());"

File: server/server_test.go [hunk 1]
```diff
@@ -1936,7 +1936,8 @@ func (cli *testServerClient) waitUntilServerOnline() {
 func (cli *testServerClient) runTestInitConnect(t *testing.T) {
 
 	cli.runTests(t, nil, func(dbt *testkit.DBTestKit) {
-		dbt.MustExec(`SET GLOBAL init_connect="insert into test.ts VALUES (NOW());SET @a=1;"`)
+		dbt.MustExec(`SET GLOBAL init_connect="insert into test.ts VALUES (NOW());`)
+		dbt.MustExec(`SET @a=1;"`)
```
Comment (djshow832) [inline server/server_test.go:1940]:
```suggestion
		dbt.MustExec(`SET GLOBAL init_connect="insert into test.ts VALUES (NOW());"`)
		dbt.MustExec(`SET @a=1`)
```

**Comment by bb7133** [comment]
I suspect there is something wrong with the parallel execution of test case:

```
> commands out of sync. Did you run multiple statements at once?
```

Which looks similiar to https://github.com/pingcap/tidb/issues/30286

**Comment by tiancaiamao** [comment]
The change seems to be incorrect.

The sql should be written as `SET GLOBAL init_connect="insert into test.ts VALUES (NOW());SET @a=1;"`
This is a feature introduced in https://github.com/pingcap/tidb/pull/23713, tidb will check the init_connect variable and execute the corresponding SQL for new incomming connection.

And the PR is merged in Mar 31 ... for a long I never see such fail in CI,
so I believe it might be related to the restruct of the test case https://github.com/pingcap/tidb/pull/29535/

There isn't much information in the error log, if it happens again I'll take a close look.

**Comment by djshow832** [comment]
> TestInitConnect

Agreed. Before migrating to testify, both `TestGetIndexMVCC` and `TestInitConnect` ran serially, but now they run in parallel and this kind of error happens frequently.

**Comment by djshow832** [comment]
> I suspect there is something wrong with the parallel execution of test case:
> 
> ```
> > commands out of sync. Did you run multiple statements at once?
> ```
> 
> Which looks similiar to #30286

Agreed. Before migrating to testify, both `TestGetIndexMVCC` and `TestInitConnect` ran serially, but now they run in parallel and this kind of error happens frequently.

**Comment by bb7133** [comment]
> Agreed. Before migrating to testify, both TestGetIndexMVCC and TestInitConnect ran serially.

Looks that we have to change them back.

**Comment by bb7133** [comment]
@tiancaiamao Could you try to clean up all the wrongly added `Parallel()` functions like:

https://github.com/pingcap/tidb/pull/29535/files#diff-40acdfe302a05645257d8fc93ba27dbecaeb828d2477fd44ff9ec8d89a5c9d6bR1101

**Comment by djshow832** [comment]
But I still have not figured out why running in parallel could cause this problem. Each `Server` runs separately with different ports in each `tidbTestSuite` and should not disturb other servers. :-(

**Comment by mjonss** [comment]
Could it be that the error really is 'bad connection'? I.e. there were some issue when connecting, which did not get handled properly?
```
[2021-11-30T02:29:26.743Z]              Error:          Received unexpected error:
[2021-11-30T02:29:26.743Z]                              bad connection
```

**Comment by mjonss** [comment]
I would suggest to add a `db.Ping()` right after the `sql.Open()` in runTests, to be sure the connection is working (Open does only open/create the handle, it does not actually connect yet).

**Comment by djshow832** [comment]
There is already a PR https://github.com/pingcap/tidb/pull/30306

**Pull request closed without merging**